orderbook/src/price_estimate.rs

```diff
@@ -1,56 +1,155 @@
+        sell_token: H160,
+        buy_token: H160,
+        amount: U256,
+        kind: OrderKind,
+    ) -> Result<f64>;
 }
 
 pub struct UniswapPriceEstimator {
     pool_fetcher: Box<dyn PoolFetching>,
+    base_tokens: HashSet<H160>,
 }
 
 impl UniswapPriceEstimator {
-    pub fn new(pool_fetcher: Box<dyn PoolFetching>) -> Self {
-        Self { pool_fetcher }
+    pub fn new(pool_fetcher: Box<dyn PoolFetching>, base_tokens: HashSet<H160>) -> Self {
+        Self {
+            pool_fetcher,
+            base_tokens,
+        }
     }
 }
 
 #[async_trait::async_trait]
 impl PriceEstimating for UniswapPriceEstimator {
-    // Estimates the price using the direct pool between sell and buy token. Price is given in
-    // how much of sell_token needs to be sold for one buy_token.
+    // Estimates the price using the direct pool between sell and buy token.
     // Returns an error if no pool exists between sell and buy token.
-    async fn estimate_price(&self, sell_token: H160, buy_token: H160) -> Result<f64> {
-        let pair = match TokenPair::new(sell_token, buy_token) {
-            Some(pair) => pair,
-            None => return Ok(1.0),
-        };
-        let pool = self
+    async fn estimate_price(
+        &self,
+        sell_token: H160,
+        buy_token: H160,
+        amount: U256,
+        kind: OrderKind,
+    ) -> Result<f64> {
+        if sell_token == buy_token {
+            return Ok(1.0);
+        }
+        let amount = U256::max(amount, U256::one());
+
+        match kind {
+            OrderKind::Buy => {
+                let (_, sell_amount) = self
+                    .best_execution_buy_order(sell_token, buy_token, amount)
+                    .await?;
+                Ok(sell_amount.to_f64_lossy() / amount.to_f64_lossy())
+            }
+            OrderKind::Sell => {
+                let (_, buy_amount) = self
+                    .best_execution_sell_order(sell_token, buy_token, amount)
+                    .await?;
+                Ok(amount.to_f64_lossy() / buy_amount.to_f64_lossy())
+            }
+        }
+    }
+}
+
+impl UniswapPriceEstimator {
+    pub async fn best_execution_sell_order(
+        &self,
+        sell_token: H160,
+        buy_token: H160,
+        sell_amount: U256,
+    ) -> Result<(Vec<H160>, U256)> {
+        self.best_execution(
+            sell_token,
+            buy_token,
+            sell_amount,
+            estimate_buy_amount,
+            estimate_buy_amount,
+        )
+        .await
+    }
+
+    pub async fn best_execution_buy_order(
+        &self,
+        sell_token: H160,
+        buy_token: H160,
+        buy_amount: U256,
+    ) -> Result<(Vec<H160>, U256)> {
+        self.best_execution(
+            sell_token,
+            buy_token,
+            buy_amount,
+            |amount, path, pools| Reverse(estimate_sell_amount(amount, path, pools)),
+            estimate_sell_amount,
+        )
+        .await
+    }
+
+    async fn best_execution<A, C, O>(
```

Comment:
Is it common in Rust to use single letter generic types? (I assume are generic types, not that far yet in the Rust book).

It's a bit hard to read IMO

Reply:
Technically it could be any CamelCase, but in practice you see mostly one letter types. Here is a somewhat related discussion: https://users.rust-lang.org/t/one-character-generics/12317

I'm happy to rename it.

Reply:
I see there's no consensus on any particular approach.

I personally prefer slightly more descriptive names when the types are not immediately clear (like when a single generic type `T`, or key value types `K, V`).

But that's my outsider view of someone who's not familiar with the code base nor the language.